### PR TITLE
Additional handler

### PR DESCRIPTION
**Pull request by EphraimB**

### CODE REVIEW

File: go-testing-tutorial/main_test.go [hunk 1]
```diff
@@ -138,3 +138,40 @@ func TestHttp(t *testing.T) {
...
 	}
 }
+
+func TestSearchHandler(t *testing.T) {
+	handler := func(w http.ResponseWriter, r *http.Request) {
+		// here we write our expected response, in this case, we return a
+		// JSON string which is typical when dealing with REST APIs
+		io.WriteString(w, "{ \"status\": \"expected service response\"}")
+
+		param1 := r.URL.Query().Get("search")
```
Comment (sherzberg) [inline go-testing-tutorial/main_test.go]: This code should live in the `searchHandler`

File: go-testing-tutorial/main_test.go [hunk 2]
```diff
@@ -138,3 +138,40 @@ func TestHttp(t *testing.T) {
...
+	if err != nil {
+		return
+	}
+	w := httptest.NewRecorder()
+	handler(w, req)
+
+	resp := w.Result()
+	body, _ := ioutil.ReadAll(resp.Body)
+
+	if resp.StatusCode == 400 {
```
Comment (sherzberg) [inline go-testing-tutorial/main_test.go]: Above you supply a `search` query string parameter, so I this check should be 200, or if you meant to test for a 400, you need to remove the `search` query string parameter above (the handler code also looks backwards from what you want).

File: go-testing-tutorial/main_test.go [hunk 3]
```diff
@@ -138,3 +138,40 @@ func TestHttp(t *testing.T) {
 		t.Error("Test failed. Expected status to equal expected service response.")
 	}
 }
+
+func TestSearchHandler(t *testing.T) {
```
Comment (sherzberg) [inline go-testing-tutorial/main_test.go]: I would change this test method name to something more specific. Maybe `TestSearchHandlerShouldReturn404IfNoSearchQueryIsPresent`.

File: go-testing-tutorial/main_test.go [hunk 4]
```diff
@@ -138,3 +138,31 @@ func TestHttp(t *testing.T) {
...
+	handler := func(w http.ResponseWriter, r *http.Request) {
+
+	}
+
+	req, err := http.NewRequest("GET", "/", nil)
+	if err != nil {
+		return
+	}
+	w := httptest.NewRecorder()
+	handler(w, req)
```
Comment (sherzberg) [inline go-testing-tutorial/main_test.go]: This doesn't test your `searchHandler`. Replace `handler` with `searchHandler`

File: go-testing-tutorial/main_test.go (hunk 2 above)
Comment (sherzberg) [reply go-testing-tutorial/main_test.go]: The logic is still backwards here. This test is supposed to fail if the response status code is not 400 (given the lack of `search` parameter)

File: go-testing-tutorial/main_test.go [hunk 5]
```diff
@@ -138,3 +139,24 @@ func TestHttp(t *testing.T) {
...
+
+	resp := w.Result()
+	body, _ := ioutil.ReadAll(resp.Body)
+
+	if resp.StatusCode != 400 {
+		t.Error("Wrong status code")
+	}
+	fmt.Println(string(body))
+
+	if string(body) != "{\"status\": \"expected service response\"}{\"jsonData\": \"{\"Results\":[\"Cutie\",\"Autism\",\"iPhone 12\"]}\"}" {
```
Comment (sherzberg) [inline go-testing-tutorial/main_test.go]: I think the response body should only be `"{\"Results\":[\"Cutie\",\"Autism\",\"iPhone 12\"]}`, no need to wrap this with `jsonData` and we dont need the additional json object that has the `status` in there.

File: go-testing-tutorial/main_test.go [hunk 6]
```diff
@@ -138,3 +139,24 @@ func TestHttp(t *testing.T) {
...
+	if err != nil {
+		return
+	}
+	w := httptest.NewRecorder()
+	searchHandler(w, req)
+
+	resp := w.Result()
+	body, _ := ioutil.ReadAll(resp.Body)
+
+	if resp.StatusCode != 400 {
```
Comment (sherzberg) [inline go-testing-tutorial/main_test.go]: This is the wrong status code to check for in this case. I believe you want to check for 200.

File: go-testing-tutorial/main.go [hunk 7]
```diff
@@ -28,6 +32,35 @@ func multiply(x int, y int) (result int) {
...
+	}
+
+	var jsonData []byte
+	jsonData, err := json.Marshal(searchResults)
+	if err != nil {
+		log.Println(err)
+	}
+	fmt.Println(string(jsonData))
+
+	io.WriteString(w, "{\"jsonData\": \""+string(jsonData)+"\"}")
```
Comment (sherzberg) [inline go-testing-tutorial/main.go]: We only want to write this data if we get a `search` query string parameter.

**Review by sherzberg — CHANGES_REQUESTED** [review]

File: go-testing-tutorial/main_test.go (hunk 6 above)
Comment (EphraimB) [reply go-testing-tutorial/main_test.go]: I should add a string query to make the status code 200?

**Review by sherzberg — APPROVED** [review]

**Pull request merged**